examples/gno.land/p/moul/authz/authz.gno

```diff
@@ -0,0 +1,252 @@
+type Authority interface {
+	// Authorize executes a privileged action if the caller is authorized
+	Authorize(title string, action PrivilegedAction) error
+
+	// String returns a human-readable description of the authority
+	String() string
+}
+
+// PrivilegedAction defines a function that performs a privileged action.
+type PrivilegedAction func() error
+
+// PrivilegedActionHandler is called by contract-based authorities to handle
+// privileged actions.
+type PrivilegedActionHandler func(title string, action PrivilegedAction) error
+
+// New creates a new Authorizer with the current realm's address as authority
+func New() *Authorizer {
+	return &Authorizer{
+		current: NewMemberAuthority(std.PrevRealm().Addr()),
+	}
+}
+
+// NewWithAuthority creates a new Authorizer with a specific authority
+func NewWithAuthority(authority Authority) *Authorizer {
+	return &Authorizer{
+		current: authority,
+	}
+}
+
+// Current returns the current authority implementation
+func (a *Authorizer) Current() Authority {
+	return a.current
+}
+
+// Transfer changes the current authority after validation
+func (a *Authorizer) Transfer(newAuthority Authority) error {
+	// Ask current authority to validate the transfer
+	return a.current.Authorize("transfer_authority", func() error {
+		a.current = newAuthority
+		return nil
+	})
+}
+
+// Do executes a privileged action through the current authority
+func (a *Authorizer) Do(title string, action PrivilegedAction) error {
+	return a.current.Authorize(title, action)
+}
+
+// String returns a string representation of the current authority
+func (a *Authorizer) String() string {
+	return a.current.String()
+}
+
+// MemberAuthority is the default implementation using addrset for member management
+type MemberAuthority struct {
+	members addrset.Set
+}
+
+func NewMemberAuthority(members ...std.Address) *MemberAuthority {
+	auth := &MemberAuthority{}
+	for _, addr := range members {
+		auth.members.Add(addr)
+	}
+	return auth
+}
+
+func (a *MemberAuthority) Authorize(title string, action PrivilegedAction) error {
+	caller := std.PrevRealm().Addr()
+	if !a.members.Has(caller) {
+		return errors.New("unauthorized")
+	}
+
+	if err := action(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (a *MemberAuthority) String() string {
+	return ufmt.Sprintf("member_authority[size=%d]", a.members.Size())
+}
+
+// AddMember adds a new member to the authority
+func (a *MemberAuthority) AddMember(addr std.Address) error {
+	return a.Authorize("add_member", func() error {
+		a.members.Add(addr)
+		return nil
+	})
+}
+
+// RemoveMember removes a member from the authority
+func (a *MemberAuthority) RemoveMember(addr std.Address) error {
+	return a.Authorize("remove_member", func() error {
+		a.members.Remove(addr)
+		return nil
+	})
+}
+
+// ListMembers returns the list of all member addresses
+// This is a read-only operation that can be called by anyone
+func (a *MemberAuthority) ListMembers() []std.Address {
```

Comment:
This can be dangerous; having this exposed could lead to unexpected behaviors and bad usage.

Reply:
This could lead to bricked or stuck usages. 👍 

Replacing it with a `rotree` getter of the underlying `avl.Tree` of `addrset`, allowing lib users to use iterators, pagers, and more.

8da98a6d223f84c0ee5e9aa69544acdd25ec236b

Reply:
I would be happy to have this `ReadOnly` helper directly on addrset  :)